### PR TITLE
[ENH] Recursive reduction using Monte-Carlo to probabilistic tabular regressors, `MCRecursiveProbaReductionForecaster`

### LINKED ISSUE #9241: [ENH] Monte Carlo sampling-based probabilistic recursive reduction forecasters using probabilistic tabular regressors
**Is your feature request related to a problem? Please describe.**
The concept is fully explored in [this related issue in skpro](https://github.com/sktime/skpro/discussions/185). This particular issue looks only at the first solution proposed by @fkiraly using Monte Carlo sampling from the predictive distributions

**Describe the solution you'd like**
A new class that holds the meta-forecaster - could be called `MCRecursiveProbaReductionForecaster` - which is an extension of the current recursive reduction forecaster but where the tabular estimator is instead a tabular probabilistic regressor and the multi-horizon probabilistic distributions are obtained by ancestral sampling.

**Describe alternatives you've considered**
Other alternatives are described in the original `skpro` issue.

### PR DESCRIPTION
**Pull request by marrov**
#### Reference Issues/PRs
Closes #9241 

#### What does this implement/fix? Explain your changes.
Adds `MCRecursiveProbaReductionForecaster` a reduction forecaster that generates multi-horizon probabilistic forecasts by recursively sampling from the probabilistic distributions of it's wrapped probabilistic tabular regressor.

Additionally:
- Adds a `BaseProbaForecaster` forecaster class to isolate the extensions done to `BaseForecaster` in order to allow for vectorized predict_proba support. My guess is that, if this is properly implemented, it could be added to the base rather than use a new base class. But I'll like to hear the reviewer's opinions on this.
- ~~Adds a throw-away class `DirectProbaReductionForecaster` that holds my initial naive (and incorrect!) approach at doing reduction with proba tabular estimators. However, it served as a good and clean first implementation and I thought I may be work keeping it at least during the review/draft period.~~
- ~~Adds a utility plotting function `plot_quantile_intervals` and `plot_distribution` which were quite useful during debugging and making scripts to showcase how to use the reducer.~~

Edit: streamlined the PR only to things directly related to `MCRecursiveProbaReductionForecaster`.

#### What should a reviewer concentrate their feedback on?

- ~~Is the cache system over engineered?~~ -> It certainly was
- Are there any issues in the main functions `_generate_mc_trajectories` and `_build_empirical_from_trajectories`? Should they be split up to shorten them?
- In `_concat_distributions_hierarchical` I'm kind of partially solving concatenation of distributions, as this is not yet present in `skpro` (see related issue: https://github.com/sktime/skpro/issues/340). If this is proper, it should be implemented there instead and removed from this forecaster.
- ~~The storing of the trajectories seems a bit contrived. I'm storing different key-value pairs of each hierarchy level (tuple) and trajectories (np array), where the level is `None` if non-multiindex. Any thoughts on how to make this more user-friendly?~~ -> Updated the storage of trajectories.

Right now the implementation works for me locally, including using hierarchical `y` and even global with exogenous data. I can provide a notebook-style example if it is useful. Below is the result of plotting the quantile intervals for the code snippet in the docstrings:

image

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice!

The failures look genuine and related to the new added forecaster.

**Comment by marrov** [comment]
> The failures look genuine and related to the new added forecaster.

@fkiraly - Yes, there was one failure due the test param handling when `skpro` was not installed so I committed a fix for this. After fixing them, some others came up that pointed to now-fixed issues in the cache handling, the trajectory handling in multiindex dfs, tags, and the handling of X in an edge case. Now all (relevant) tests are passing!

Regarding the ones I'm seeing now  in `01c_forecasting_hierarchical_global`, seem to be related to torch, so nothing to do with my PR as it shows `KeyError: 'weights_only'`. ~~I can try to pull in the latest changes in case there was some recent fix for this.~~ 
- Edit 1: I tried to pull latest changes, but the error is still there. I believe this should not be caused by my PR.
- Edit 2: I am now really really sure this is not my PR as recent an unrelated PR #9270 also has the same error, see [this CICD run](https://github.com/sktime/sktime/actions/runs/21170910405/job/60887222490?pr=9270).

**Comment by felipeangelimvieira** [comment]
Are the current failures related to the changes?
The changes look good to me. After updating the existing RRF, we should probably also revisit this implementation to check if it can be more efficient? How fast is this current implementation?

**Comment by marrov** [comment]
> Are the current failures related to the changes? 

I do not think so, no. 

> After updating the existing RRF, we should probably also revisit this implementation to check if it can be more efficient? How fast is this current implementation?

I'd say it's pretty decent already. The fit-predict loop with a df of hierarchical 4 series each with 1000 time points generating 50,000 MC samples for the distributions takes under 10 seconds using `xgboostlss`. A plot from one of these series looks like this:

image

However, it's probably not difficult to find ways to speed up the core algorithms.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Really great!

I think the concat for the distribution should be in the base class, in general. Have you checked whether it works for everything else?

I would suggest we do that and then merge this.

Additionally, I would later try to move `_concat_distributions_hierarchical` to a new module `skpro.distributions.concat`, although I am not sure how to solve best that `VectorizedDF` does not exist there.
I would do this only after this PR is merged though.

**Comment by marrov** [comment]
@fkiraly - Move concat to the base class. I also did some general cleanup & refactoring, added test coverage, and handled edge cases (main one being in-sample fh). 

Should be ready for final review and merging.

> Have you checked whether it works for everything else?

No, just checked it for `Empirical`.

**Comment by marrov** [comment]
Not sure why the copilot review triggered twice but it did give some decent suggestions that I implemented. I also run a local code review and got a suggestion for a vectorization performance improvement (fyi @felipeangelimvieira ) that was implemented. All CI tests are passing and the PR is ready for review @fkiraly .

**Review by fkiraly — CHANGES_REQUESTED** [review]
Approved in-principle.

blocking: the estimator should be added to the API reference for forecasters, in the "reduction" section.

Minor comments, non-blocking:

* I hope ultimately you are planning to move the vectorization capability to `skpro`?
* using the lagger as in the 2nd generation reducer has proven to introduce computational inefficiencies - @ericjb has looked at this. Perhaps at some point performance optimizations could be done, but this is not necessary right now.

**Comment by marrov** [comment]
> blocking: the estimator should be added to the API reference for forecasters, in the "reduction" section.

Done, thanks for the reminder.


> * I hope ultimately you are planning to move the vectorization capability to `skpro`?

Yes, that is the plan.

> * using the lagger as in the 2nd generation reducer has proven to introduce computational inefficiencies - @ericjb has looked at this. Perhaps at some point performance optimizations could be done, but this is not necessary right now. 

Interesting! I also saw a comment from @felipeangelimvieira in the v1 issue about a possible rework of the reducers. Maybe that could also have some nice learnings for this forecaster.

**Comment by marrov** [comment]
@felipeangelimvieira and @benHeid  - You mentioned on discord that you'd take a look at the PR. Still got time for it? I'd stil. love to get your thoughts but if you don't have the bandwidth maybe we can go ahead.

**Comment by benHeid** [comment]
I am on vacation until tomorrow evening. So this weekend I can take a look at this PR. But feel free to go ahead if you prefer to finalize it soon.

File: sktime/forecasting/compose/_reduce_proba.py [hunk 1]
```diff
@@ -0,0 +1,784 @@
...
+            exog_cols = pd.Index(X_pool.columns).drop_duplicates()
+            X_fallback_df = _get_last_X_for_index(X_pool, x_template_index)
+            X_fallback_df = _align_X_columns(X_fallback_df, exog_cols)
+            try:
+                X_at_idx = slice_at_ix(X_pool, first_predict_idx)
+                if len(X_at_idx) == 0:
+                    raise KeyError(first_predict_idx)
+                X_at_idx = _align_X_index(X_at_idx, x_template_index)
+                X_at_idx = _align_X_columns(X_at_idx, exog_cols)
+                X_at_idx = X_at_idx.fillna(X_fallback_df)
```
Comment (benHeid) [inline sktime/forecasting/compose/_reduce_proba.py]: Can `X_fallback_df` be None?  In line 539 you are checking directly before more or less the same statement.

File: sktime/forecasting/compose/_reduce_proba.py [hunk 2]
```diff
@@ -0,0 +1,784 @@
...
+        X_features_by_step = {}
+        if X_pool is not None:
+            for step_idx in range(max_horizon):
+                horizon = step_idx + 1
+                fh_step = ForecastingHorizon(
+                    [horizon], is_relative=True, freq=self._cutoff
+                )
+                fh_step_abs = fh_step.to_absolute_index(self._cutoff)
+                predict_time = fh_step_abs[0]
+                try:
```
Comment (benHeid) [inline sktime/forecasting/compose/_reduce_proba.py]: This block looks very similar to the block in 501ff?

Maybe, you could extract this in a separate method.

**Review by benHeid — COMMENTED** [review]
Personally, I struggled to understand all parts of the algorithm since it is a lot of code per method. Especially in `_generate_mc_trajectories`. Maybe extracting some parts of the code in separate methods might improve the understandability. 

Based on this, I just comment on the review. Feel free to merge this nevertheless, since Franz approved this PR in principle already.

**Comment by marrov** [comment]
> Personally, I struggled to understand all parts of the algorithm since it is a lot of code per method. Especially in `_generate_mc_trajectories`. Maybe extracting some parts of the code in separate methods might improve the understandability.

Thanks for the review! I agree the code gets quite heavy at times. I've made a follow-up commit with a small refactor for `_generate_mc_trajectories` to factor out some methods. It does not fully address the issue with the algo being long, but hope it dells. I believe `_get_exog_at_time` should address both of your specific comments.

Let me know if you have any further thoughts, otherwise I think this should be in a "mergeable" state now @fkiraly.

**Review by fkiraly — DISMISSED** [review]

**Review by fkiraly — DISMISSED** [review]
Something is going on in `test_multivariate_y_smoke` - I think the multivariate concatenation is not properly working. Change in `pandas` API perhaps? Feels like an argument to ensure this is in `skpro` and properly tested there...

**Comment by marrov** [comment]
@fkiraly - Found it. The issue was caused by pulling in #9212, which deprecates `scitype:y`. Apparently, after that change the alias resolution in `TagAliaserMixin` breaks when querying the old `capability:multivariate` because it picks univariate-only as the first matching old tag (not `scitype:y`). This happens because of this line:
```python
old_tag_name = [k for k, v in alias_dict.items() if v == tag_name][0]
```
Then as it finds it absent, it falls back to `False`. So even though `scitype:y: "both"` was set, it was silently ignored and the forecaster got column-vectorized on multivariate y, which broke `predict_proba`. Fun bug.

Good thing is that the fix is quite easy, just to drop `scitype:y` and use `capability:multivariate: True` instead.

**Comment by marrov** [comment]
@fkiraly , @benHeid - Should be ok to merge, made some updates based on CI errors on unequal length series. Currently only docs build is failing, which seems to be the case everywhere.

**Review by fkiraly — APPROVED** [review]
